[Remote Clusters] Migrate to the New Platform with a shim

cjcenizal opened the pull request:
This PR also makes some changes to `createRouter`, namely by exporting helpers for creating errors.

elasticmachine commented:
Pinging @elastic/es-ui

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/16301/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/16310/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/16339/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/16699/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/16702/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/16722/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/16765/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/16805/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/16918/)

cjcenizal commented:
Retest

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/16976/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/17031/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/17044/)

cjcenizal commented on x-pack/server/lib/create_router/index.js:
```diff
@@ -9,8 +9,18 @@ import { isEsErrorFactory } from './is_es_error_factory';
+export { wrapEsError, wrapUnknownError, wrapCustomError } from './error_wrappers';
+
+// Sometimes consumers will need to check if errors are ES errors, too.
+export const createIsEsError = server => {
```
I think we'll want the ability to easily hand-roll our own errors when using `createRouter`, so to me it makes sense to export these helpers from the same module.

jen-huang replied:
can we just export `isEsErrorFactory` instead?
`export { isEsErrorFactory } from './is_es_error_factory';`
it is a little weird to export and import the same module in the same file though 🙂

cjcenizal commented on x-pack/test/api_integration/apis/management/remote_clusters/remote_clusters.js:
```diff
@@ -95,18 +90,6 @@ export default function ({ getService }) {
         });
       });
-
-      it(`fails if seeds aren't provided`, async () => {
```
This caused problems because the `updateHandler` implementation deletes the cluster before attempting to update it, which then fails in this test case. This causes the `List` and `Delete` tests to fail. I'm not sure why this wasn't failing consistently though.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/73/)

cjcenizal commented on x-pack/plugins/remote_clusters/index.ts:
```diff
@@ -0,0 +1,68 @@
+      injectDefaultVars(server: Legacy.Server) {
+        const config = server.config();
+        return {
+          remoteClustersUiEnabled: config.get('xpack.remote_clusters.ui.enabled'),
```
Works

elasticmachine commented:
Pinging @elastic/kibana-platform

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/103/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/120/)

joshdover commented:
Today is looking pretty full, but I'll get to this before the EOW.

joshdover commented on x-pack/plugins/remote_clusters/public/shim.js:
```diff
@@ -0,0 +1,52 @@
+      },
+      documentation: {
+        esDocBasePath: `${ELASTIC_WEBSITE_URL}guide/en/elasticsearch/reference/${DOC_LINK_VERSION}/`,
+      },
```
The shape of this is a pretty different than CoreStart. It's definitely helpful to use TypeScript here to have it guide the correct shape.

This might be a bit closer:
```js
return {
  core: {
    i18n: { Context: I18nContext },
    chrome: {
      // Only include chrome APIs we actually support in CoreStart.
      // It's actually like that you don't need chrome at all and instead need `http.basePath` for example.
    },
    notification: {
      toast: toastNotifications, // I'm not sure, but I think this API is actually slightly different from CoreStart.
    },
    http: {
      basePath: { prepend }
    }
  }
}
```

You'll notice I removed some things that are not present CoreStart (and likely won't ever be). I think it's best to do this now (or in a subsequent PR if you prefer) so that you can start shaping your application the correct way.

For instance, the `FormattedMessage` component should be imported directly from @kbn/i18n rather than injected via core contracts. Another thing that core will not support is routing. We recommend you setup ReactRouter similar to how APM is doing this.

If you want to do this more incrementally, you can start with what you have here and start replacing it with pieces of the real CoreStart instance from `import { npStart: { core } } from 'ui/new_platform'` in future PRs, rather than constructing your own.

It's also worth mentioning that this interface is changing a bit right now. For instance, Chrome and I18n are moving from CoreSetup to CoreStart in #37915

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1457/)

cjcenizal commented on x-pack/plugins/remote_clusters/public/shim.ts:
```diff
@@ -0,0 +1,52 @@
+      },
+      routes: {
+        register: routes.when.bind(routes),
+      },
```
@epixa What's the best way for teams to submit requests/proposals for additions or changes to the core API? For example, here I add `fatalError`, `documentation`, and `routes` to the interface. I'd like to bring these suggestions to the awareness of the NP team and other engineers participating in the migration in a central, visible location.

epixa replied:
Issues or for sufficiently complicated additions an RFC PR. And one of the purposes of the platform consultant thing is so you can just offload that sort of thing directly on that person.

cjcenizal commented:
@jen-huang This PR is ready for your 👀 !

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1513/)

jen-huang commented on x-pack/plugins/remote_clusters/plugin.ts:
```diff
@@ -0,0 +1,30 @@
+    registerGetRoute(router);
+    registerAddRoute(router);
+    registerUpdateRoute(router);
+    registerDeleteRoute(router, isEsError);
```
I don't love that this method signature is different than the other ones. I would prefer one `registerRoutes()` method call that calls all the other ones [like in SR](https://github.com/elastic/kibana/blob/master/x-pack/plugins/snapshot_restore/server/routes/api/register_routes.ts). any additional helpers can then be passed as part of `Plugins` interface, or maybe a `Core` interface in this case.

jen-huang commented on x-pack/plugins/remote_clusters/public/app/services/notification.ts:
```diff
@@ -4,4 +4,10 @@
+export let fatalError: any;
+
+export function init(_toasts: any, _fatalError: any): void {
+  toasts = _toasts;
```
my preference is to keep this name as `toastNotifications` so that this app's usages of toasts show up when grepping for all `toastNotifications` references in the codebase

cjcenizal commented on x-pack/plugins/remote_clusters/public/shim.ts:
```diff
@@ -0,0 +1,52 @@
+      },
+    },
+    pluginsStart: {
+      management: {
```
@lukeelmers We've shimmed the Management plugin's `start` contract. I thought you might be interested in what we depend upon and/or possibly want to provide us some feedback.

cjcenizal commented:
Thanks for your feedback @jen-huang!

> can we just export isEsErrorFactory instead?

Sure, will do!

> I don't love that this method signature is different than the other ones. I would prefer one registerRoutes() method call that calls all the other ones like in SR. any additional helpers can then be passed as part of Plugins interface, or maybe a Core interface in this case.

Hmm, I think we're approaching this in different ways. I'm trying to apply the [Law of Demeter](https://en.wikipedia.org/wiki/Law_of_Demeter) by making each method depend upon specific services instead of depending upon the service locators (`Core` and `Plugins`). To me, the biggest advantage to this is that only one file depends on the structure of these service locators, so we only have one place to make a change if they change their structure.

In terms of actors and relationships, the big difference I see between the SR code and this code is the extra level of indirection created by the `registerRoutes` method. I removed it here because I didn't feel like it offered a helpful abstraction. To me, making `plugin.ts` responsible for registering the routes made the role of that file clearer and simplified the code by removing one level of indirection. I might not be fully appreciating the code from your perspective though, so I'd be happy to Zoom about this!

> my preference is to keep this name as toastNotifications so that this app's usages of toasts show up when grepping for all toastNotifications references in the codebase

I don't feel strongly about this, so I'm open to changing it back to `toastNotifications`, but the reason I made this change is because the [coreStart API](https://github.com/elastic/kibana/blob/master/docs/development/core/public/kibana-plugin-public.notificationsstart.md) calls them `toasts` and as everyone migrates to the New Platform we'll see other apps use `toasts` instead of `toastNotifications` too. To grep for all occurrences of both, we can use the regex `(toastNotifications\.)|(toasts\.)`. With this in mind, do you still think we should go back to using `toastNotifications`?

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1538/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1545/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1560/)

joshdover commented on x-pack/plugins/remote_clusters/public/shim.ts:
```diff
@@ -21,19 +20,14 @@ export function createShim() {
-        fatalError,
-      },
+      notifications,
+      fatalError,
```
Just realized I was wrong and fatalError is already exposed, but right now it's only on CoreSetup. I'll put up a PR for exposing it in CoreStart as well, but it should be the exact same interface as CoreSetup.

Instead of getting this from `ui/notify` you should be able to get it from `npSetup.core.fatalErrors`. This is important because the interface in `ui/notify` is slightly different than the one exposed by Core (but should be a very easy refactor).

cjcenizal replied:
Great, thanks!

joshdover replied:
Been thinking about this more, and I'm not sure it makes sense to support this in the new platform. In the new platform:

- If a plugin fails to setup or start, they can throw exceptions during their `setup` and/or `start` method and Core will automatically display that error in the full-page error UI.
- If an application fails to mount or throws some error in the UI, that app should handle it with an error boundary React component to display the error without taking down the entire frontend.

In the case of your migration, it may make sense to use `npSetup.core.fatalError` in a `catch` block around `start` in your shim and replace any other usages with a React error boundary component that also calls `npSetup.core.fatalError` for now.

Thoughts?

cjcenizal replied:
> If an application fails to mount or throws some error in the UI, that app should handle it with an error boundary React component to display the error without taking down the entire frontend.

I agree with this principle. In terms of the ES UI apps, `fatalError` is only used in a dozen places so migrating off it won't be difficult. I wonder if the Platform team can ease the migration process by wrapping the mountpoint in an `EuiErrorBoundary` or framework-agnostic equivalent (maybe even a `failApp` method that can be used in lieu of `fatalError`). The idea being that an app developer has a mechanism for "bailing out" with a descriptive error message if something absolutely unexpected occurred, without needing to implement this themselves. Using this mechanism would display an error in place of the app, without taking down all of Kibana. This will also ensure that there's a visually consistent way of presenting this information so each app doesn't look different to the user when it fails in this manner.

joshdover replied:
I agree, consistency here is ideal. I'll add that as an additional to the ApplicationService project. In the meantime, I think it's fine to go ahead and keep using `fatalError` until we have another solution.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1732/)

cjcenizal commented:
@jen-huang Thanks! FWIW, I think logically grouping many endpoints the way you did in SR is great. It sounds like it's just a matter of preference, but I would have less mental overhead if the dependencies were passed in like this because then I can tell what each set of routes really needs:

```js
export class Plugin {
  public start(core: Core, plugins: Plugins): void {
    const router = core.http.createRouter(API_BASE_PATH);
    const { cloud, elasticsearch, xpack_main: xPackMain } = plugins;

    // Register routes
    registerAppRoutes(router, xPackMain);
    registerRepositoriesRoutes(router, elasticsearch, cloud);
    registerSnapshotsRoutes(router);
  }
}
```
